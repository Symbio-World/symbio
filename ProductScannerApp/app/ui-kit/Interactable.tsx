import * as React from 'react'
import { ViewStyle } from 'react-native'
import { spring, snapPoint } from 'react-native-redash'
import { PanGestureHandler, State } from 'react-native-gesture-handler'
import Animated from 'react-native-reanimated'

const {
  Value,
  event,
  block,
  set,
  cond,
  eq,
  Clock,
  call,
  clockRunning,
  interpolate,
} = Animated

const binaryInterpolation = (
  value: Animated.Adaptable<number>,
  origin: Animated.Adaptable<number>,
  destination: Animated.Adaptable<number>,
) =>
  interpolate(value, {
    inputRange: [0, 1],
    outputRange: [origin, destination],
  })

type Value = typeof Value

type SnapPoint = {
  x: number
}

type InteractableProps = {
  x: Animated.Value<number>
  y: Animated.Value<number>
  snapPoints: SnapPoint[]
  onSnap: (e: { nativeEvent: { x: number } }) => void
  style: ViewStyle
}

export const Interactable: React.FC<InteractableProps> = ({
  style,
  x,
  y,
  snapPoints,
  onSnap,
  children,
}) => {
  const clock = new Clock()
  const springValue = new Value(0)
  const translationX = new Value(0)
  const translationY = new Value(0)
  const velocityX = new Value(0)
  const snapPointX = new Value(0)
  const state = new Value(State.UNDETERMINED)
  const onGestureEvent = event([
    {
      nativeEvent: {
        velocityX,
        translationX,
        translationY,
        state,
      },
    },
  ])
  const points = snapPoints.map((point) => point.x)
  Animated.useCode(
    () =>
      block([
        cond(eq(state, State.END), [
          set(snapPointX, snapPoint(translationX, velocityX, points)),
          set(springValue, spring({ clock, from: 0, to: 1 })),
          cond(
            eq(clockRunning(clock), 0),
            call([snapPointX], ([lastX]) =>
              onSnap({ nativeEvent: { x: lastX } }),
            ),
          ),
        ]),
        set(
          x,
          cond(
            eq(state, State.ACTIVE),
            translationX,
            binaryInterpolation(springValue, translationX, snapPointX),
          ),
        ),
        set(
          y,
          cond(
            eq(state, State.ACTIVE),
            translationY,
            binaryInterpolation(springValue, translationY, 0),
          ),
        ),
      ]),
    [],
  )

  return (
    <PanGestureHandler
      onHandlerStateChange={onGestureEvent}
      onGestureEvent={onGestureEvent}
    >
      <Animated.View style={style}>{children}</Animated.View>
    </PanGestureHandler>
  )
}
