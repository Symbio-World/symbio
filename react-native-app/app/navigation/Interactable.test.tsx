import * as React from 'react'
import { render } from 'react-native-testing-library'
import { Text, StyleSheet, Dimensions } from 'react-native'

import { Interactable } from './Interactable'

jest.mock('react-native-reanimated', () =>
  require('react-native-reanimated/mock'),
)
import Animated, { Extrapolate } from 'react-native-reanimated'

const { width, height } = Dimensions.get('window')
const deltaX = width / 2
const α = Math.PI / 12
const A = width * Math.cos(α) + height * Math.sin(α)

describe('Interactable', () => {
  const onSnap = jest.fn()
  const x = new Animated.Value(0)
  const y = new Animated.Value(0)
  const translateX = x
  const translateY = y
  const rotateZ = Animated.concat(
    Animated.interpolate(x, {
      inputRange: [-1 * deltaX, deltaX],
      outputRange: [α, -1 * α],
      extrapolate: Extrapolate.CLAMP,
    }),
    'rad',
  )
  const style = [
    StyleSheet.absoluteFill,
    {
      transform: [{ translateX }, { translateY }, { rotateZ }],
    },
  ]

  it('renders correctly', () => {
    const { toJSON } = render(
      <Interactable
        style={style}
        snapPoints={[{ x: -1 * A }, { x: 0 }, { x: A }]}
        onSnap={onSnap}
        x={x}
        y={y}
      >
        <Text>Text</Text>
      </Interactable>,
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
