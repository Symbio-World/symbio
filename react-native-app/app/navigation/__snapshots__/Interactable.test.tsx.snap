// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactable renders correctly 1`] = `
<View
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "transform": Array [
          Object {
            "translateX": Object {
              "setValue": [Function],
            },
          },
          Object {
            "translateY": Object {
              "setValue": [Function],
            },
          },
          Object {
            "rotateZ": "undefinedrad",
          },
        ],
      },
    ]
  }
>
  <Text>
    Text
  </Text>
</View>
`;
